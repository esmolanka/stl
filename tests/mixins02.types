
type Foo = mixin
  { foo : Int
  , bar : Float
  }

type FooRecord = { mix Foo }
#eval FooRecord

type Identity a = a
type FooRecord2 = record of Identity { mix Foo }
#eval FooRecord2

type Array (n : Nat) a = a[n]
type FooTable = exists (n : Nat). record of (Array n) { mix Foo }
#eval FooTable

type Maybe a =
  | Nothing
  | Just : a
type FooOverride = exists (n : Nat). record of Maybe { mix Foo }
#eval FooOverride
