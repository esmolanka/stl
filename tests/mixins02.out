tests/mixins02.types:8:1-8:16:
  Normalised:
    ∃ (ρ : Row). {foo : Int, bar : Float, … ρ}
  Kind:
    Type
tests/mixins02.types:12:1-12:17:
  Normalised:
    ∃ (ρ : Row). {foo : Int, bar : Float, … ρ}
  Kind:
    Type
tests/mixins02.types:16:1-16:15:
  Normalised:
    ∃ (n : Nat). ∃ (ρ : Row). {foo : Int[n], bar : Float[n], … ρ}
  Kind:
    Type
tests/mixins02.types:22:1-22:18:
  Normalised:
    ∃ (n : Nat). ∃ (ρ : Row).
      { foo : ∀ (ρ : Row). <Nothing? : Unit | Just? : Int | … ρ>
      , bar : ∀ (ρ : Row). <Nothing? : Unit | Just? : Float | … ρ>
      , … ρ
      }
  Kind:
    Type